tool/tctl/common/bots_command.go

```diff
@@ -546,6 +526,296 @@ func (c *BotsCommand) UpdateBot(ctx context.Context, client *authclient.Client)
+	for _, i := range instances {
+		var (
+			joinMethod string
+			hostname   string
+			generation string
+		)
+
+		joined := i.Status.InitialAuthentication.AuthenticatedAt.AsTime().Format(time.RFC3339)
+		initialJoinMethod := i.Status.InitialAuthentication.JoinMethod
+
+		lastSeen := i.Status.InitialAuthentication.AuthenticatedAt.AsTime()
+
+		if len(i.Status.LatestAuthentications) == 0 {
+			generation = "n/a"
+		} else {
+			auth := i.Status.LatestAuthentications[len(i.Status.LatestAuthentications)-1]
+
+			generation = fmt.Sprint(auth.Generation)
+
+			if auth.JoinMethod == initialJoinMethod {
+				joinMethod = auth.JoinMethod
+			} else {
+				// If the join method changed, show the original method and latest
+				joinMethod = fmt.Sprintf("%s (%s)", auth.JoinMethod, initialJoinMethod)
+			}
+
+			if auth.AuthenticatedAt.AsTime().After(lastSeen) {
+				lastSeen = auth.AuthenticatedAt.AsTime()
+			}
+		}
+
+		if len(i.Status.LatestHeartbeats) == 0 {
+			hostname = "n/a"
+		} else {
+			hb := i.Status.LatestHeartbeats[len(i.Status.LatestHeartbeats)-1]
+
+			hostname = hb.Hostname
+
+			if hb.RecordedAt.AsTime().After(lastSeen) {
+				lastSeen = hb.RecordedAt.AsTime()
+			}
+		}
+
+		t.AddRow([]string{
+			fmt.Sprintf("%s/%s", i.Spec.BotName, i.Spec.InstanceId), joinMethod,
+			hostname, joined, lastSeen.Format(time.RFC3339), generation,
+		})
+	}
+	fmt.Fprintln(c.stdout, t.AsBuffer().String())
+
+	fmt.Fprintf(c.stdout, "\nTo view more information on a particular instance, run:\n\n> %s bots instances show [id]\n", os.Args[0])
+
+	if c.botName != "" {
+		fmt.Fprintf(c.stdout, "\nTo onboard a new instance for this bot, run:\n\n> %s bots instances add %s\n", os.Args[0], c.botName)
+	}
+
+	return nil
+}
+
+// AddBotInstance begins onboarding a new instance of an existing bot.
+func (c *BotsCommand) AddBotInstance(ctx context.Context, client *authclient.Client) error {
+	// A bit of a misnomer but makes the terminology a bit more consistent. This
+	// doesn't directly create a bot instance, but creates token that allows a
+	// bot to join, which creates a new instance.
+
+	bot, err := client.BotServiceClient().GetBot(ctx, &machineidv1pb.GetBotRequest{
+		BotName: c.botName,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var token types.ProvisionToken
+
+	if c.tokenID == "" {
+		// If there's no token specified, generate one
+		tokenName, err := utils.CryptoRandomHex(defaults.TokenLenBytes)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		ttl := c.tokenTTL
+		if ttl == 0 {
+			ttl = defaults.DefaultBotJoinTTL
+		}
+		tokenSpec := types.ProvisionTokenSpecV2{
+			Roles:      types.SystemRoles{types.RoleBot},
+			JoinMethod: types.JoinMethodToken,
+			BotName:    c.botName,
+		}
+		token, err = types.NewProvisionTokenFromSpec(tokenName, time.Now().Add(ttl), tokenSpec)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if err := client.UpsertToken(ctx, token); err != nil {
+			return trace.Wrap(err)
+		}
+	} else {
```

Comment:
Do we need the else or have all possible code paths in the if been exhausted?

Reply:
I can dedent this branch, sure.